generators/profilegen/src/resources/common/configurable.rb

```diff
@@ -0,0 +1,84 @@
+# encoding: utf-8
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+
+module Azure::Common
+  # The Azure::Common::Configurable module provides basic configuration for Azure activities.
+  module Configurable
+    # @return [String] Azure tenant id (also known as domain).
+    attr_accessor :tenant_id
+
+    # @return [String] Azure client id.
+    attr_accessor :client_id
+
+    # @return [String] Azure secret key.
+    attr_accessor :client_secret
+
+    # @return [String] Azure subscription id.
+    attr_accessor :subscription_id
+
+    # @return [MsRestAzure::ActiveDirectoryServiceSettings] Azure active directory service settings.
+    attr_accessor :active_directory_settings
+
+    # @return [MsRest::ServiceClientCredentials] credentials to authorize HTTP requests made by the service client.
+    attr_accessor :credentials
+
+    class << self
+      #
+      # List of configurable keys for {Azure::Common::Client}.
+      # @return [Array] of option keys.
+      #
+      def keys
+        @keys ||= [:tenant_id, :client_id, :client_secret, :subscription_id, :active_directory_settings]
+      end
+    end
+
+    #
+    # Set configuration options using a block.
+    #
+    def configure
+      yield self
+    end
+
+    #
+    # Resets the configurable options to provided options or defaults.
+    # This will also creates MsRest::TokenCredentials to be used for subsequent Azure Resource Manager clients.
+    #
+    def reset!(options = {})
+      Azure::Common::Configurable.keys.each do |key|
+        default_value = Azure::Common::Default.options[key]
+        instance_variable_set(:"@#{key}", options.fetch(key, default_value))
+      end
+
+      default_value = get_credentials(self.tenant_id, self.client_id, self.client_secret, self.active_directory_settings)
```

Comment:
is this needed? 
if credentials are not passed in the options, should the "fetch" above get "credentials" from Configurable and set the value to the same thing that "get_credentials" below is doing?

Reply:
I wonder why we need to get credentials in Configurable and remove it from Default, where it was.

Reply:
As mentioned in the previous comment, the user may or may not provide credentials in the options. If he provides it, we will use it. Else, we will derive it from the other values supplied. 

The 2 fetch are different. The first one (in setup_options) is getting the credentials from the default value and the second one is from the user supplied values (not the environment variables).